### ISSUE TITLE
RFC: Submissions design rework

### ISSUE THREAD

**Issue by tomaka** [tomaka/vulkano#385]
# Summary

Under this new proposal you would not just simply submit a command buffer, but submit a
command buffer *after* a previously submitted command buffer. In other words, you wouldn't just
call `cb.submit()` but `cb.exec_after(previous_submission)`.

Of course it would still be possible to submit a command buffer alone, because when a program
submits its first ever command buffer there is no previous submission. However that would be
the exception rather than the rule.

Handling resources synchronization would be done in two steps:

- First you acquire a kind of "lock" on a resource in order to start using it from the GPU, and
  pass it to either to the first command buffer of a chain or to all the command buffers of a chain.
- Then you can contine using this resource lock in all the command buffers of that chain.

# Futures

This proposal introduces a trait named `GpuFuture` which represents an event that will happen in
the future. The `Gpu` prefix is here because the GPU is usually involved, but it's not always the
case.

We'll see later the actual definition of the `GpuFuture` trait.

So which structs implement `GpuFuture`?

- When you submit a command buffer to a queue, you obtain a `CommandBufferExecFuture` struct
  that represents the moment when the queue finishes executing it.
- You can ask the GPU to signal a semaphore, in which case you obtain a `SemaphoreSignalFuture`
  struct that represents the moment when the semaphore is going to be signaled.
- You can also signal multiple semaphores at once, in which case you obtain a
  `MultiSemaphoreSignalFuture`.
- You can ask the GPU to signal a fence, in which case you obtain a `FenceSignalFuture` struct
  that represents the moment when the fence is going to be signaled.
- When you acquire an image from a swapchain, you obtain a `SwapchainAcquireFuture` that represents
  the moment when the GPU acquires the lock on the swapchain image.
- You can join multiple futures into one with a `JoinFuture`.

As explained in the summary, the idea behind this design is that most of the time you will ask the
GPU to execute things after a given point in time represented by a future.

For example if you call `cb.exec_after(sem_sig)` where `sem_sig` is a future that represents a
semaphore being signaled, then the command buffer will only start executing after the semaphore is
signaled. Behind the scenes the `exec_after` function adds the semaphore as a waiting semaphore
when it submits the command buffer to the Vulkan implementation.

# Submitting

Calling `command_buffer.exec_after(...)` would return a future that represents the moment when the
command buffer has finished executing, but doesn't actually submit the execution immediately.
Similarly you can add actions to execute after a future (eg. execute another command buffer or
signal a semaphore), but doing so doesn't immediately submit these actions.

Instead the process of submitting the future must be done manually by the user.

Therefore the idea is to build a chain of command buffers, and submit this chain from time to
time. This corresponds to what experts usually recommend doing in a Vulkan program.

Submitting a future could be done with a mutable method on said future. Each future also submits
the previous futures in the chain. Each future remembers whether it has been submitted in order
to avoid submitting the same thing twice.

Trying to submit things twice should not result in an error, as the function can get called from
other futures downstream in the chain. Therefore the function should probably be called `flush()`
rather than `submit()`, to avoid confusing users.

# Safety checks

While the futures system is a nice way to express dependencies between command buffers, its actual
main purpose is checking and enforcing the safety of the submissions.

Whenever a command buffer is submitted after a future, the resources used by the command buffer are
enumerated and the future is queried to check if it gives access to that resource.

For example let's say you submit a command buffer that writes to a buffer, and obtain a future
corresponding to the moment when the execution finishes. You then submit another command buffer
that modifies the same buffer after that future. Since we know that the first command buffer
has exclusive access to that buffer, any command buffer executed afterwards also automatically has
exclusive access to that same buffer. Vulkano queries the future, which in turn queries the command
buffer, in order to confirm that we have exclusive access.

The exact rules are:

- If we submit a command buffer that has exclusive access to a resource, the command buffer that
  comes after also has exclusive or shared access to that resource.
- An exception to this rule ^ is if we append a `MultiSemaphoreSignalFuture` which signals multiple
  semaphores. Since this future gives the possibility to submit multiple command buffers that all
  start simultaneously, each one only gets shared access.
- If we submit a command buffer that has shared access to a resource, the command buffer that
  comes after also has shared access to that resource.

If a future can't determine that it has access to a resource, it queries the future beforehand,
which in turn queries the future beforehand.

But a question remains: how to start from scratch? Let's say we initialize vulkano, create a
buffer, and submit a command buffer that modifies this buffer. Since we have nothing to submit
before, there's no way to determine whether we have an exclusive access through this mechanism
alone.

That's why we need the help of the resources themselves.

## Resources design

This RFC proposes to change the philosophy behind buffers and images. Instead of being simply
"dumb" resources that can be accessed, they will need to deeply integrate the fact that they can
be used in a chain of command buffers.

You can think of it like this: `UnsafeBuffer` and `UnsafeImage` are like a `T`, while
`CpuAccessibleBuffer`, `SwapchainImage`, `AttachmentImage`, and so on are like a `Mutex<T>`.
In other words, these types add the synchronization primitive around buffers and images. Note that
this is already a bit the case before this RFC, but this RFC proposes to push this even more.

The rules are as follow:

- As described in the previous section, when you submit a command buffer after a future, that
  future is checked first for each resource.
- If the future returns "false" for a resource, meaning that it is unknown whether the buffer/image
  can be accessed, then a method on the buffer/image is called to ask whether it is ok to access
  it.
- The buffer or the image remembers that it is in use by the GPU and stays in this state until
  destroyed. This means that you wouldn't pass actual buffers or images, but handles that point to
  a buffer or image and that implement the Buffer/Image trait.
- When the buffer/image is already in use, calling that method returns an error.

This design also solves the problem that currently exists in vulkano where resources must be locked in the correct order. Right now vulkano has a global lock for whenever it wants to submit something. This global lock would be gone.

What does it mean in practice?

For example, for `CpuAccessibleBuffer`, instead of writing this:

```rust
let buffer = Arc::new(CpuAccessibleBuffer::new(...));
CommandBufferBuilder::new().fill_buffer(buffer, 0).build().submit(...);
```

You'd write this:

```rust
let buffer = CpuAccessibleBuffer::new(...);
let buffer_for_my_frame = buffer.start().unwrap();
CommandBufferBuilder::new().fill_buffer(buffer_for_my_frame, 0).build().submit(...);
```

The `.start()` method (whose name will probably change) on the `CpuAccessibleBuffer` creates a
lightweight object (for example of type `CpuAccessibleBufferStarted`) that represents an access to
the buffer for a chain a command buffers. As long as this object is alive, calling `.start()` again
on the `CpuAccessibleBuffer` would return a runtime error. If you want to use the
`CpuAccessibleBuffer` buffer multiple times in a row, you have to chain its usages with futures.

Once the command buffers have been submitted with a fence, and the `FenceSignalFuture` has been destroyed, the `CpuAccessibleBufferStarted` handle is destroyed as well and the `CpuAccessibleBuffer` becomes usable again.

Note that this makes the `CpuAccessibleBuffer` a bit clumbsy to use. This type would probably
be deprecated and eventually removed in favor of pools that allow uploading and downloading data.

In practice the following wrappers would be available:

- `SwapchainImage` would only be accessible through a future obtained when calling
  `acquire_next_image`. It could therefore only belong to one chain at a time.
- `ImmutableImage` and `ImmutableBuffer` would have a mechanism for the initial upload, then no
  mechanism at all, since they are immutable anyway.
- `UploadBuffersPool` contains several subbuffers (three?). Calling `.next(data)` would grant
  access to one of the subbuffers. Same for `DownloadBuffersPool`.
- `CpuAccessibleBuffer`, `AttachmentImage` and `StorageImage` would have the mechanism where they
  can only belong to one chain at a time.

# Destroying futures

The other job of futures is to keep command buffers and resources alive while they are being
accessed by the GPU. Therefore destroying a future would trigger a panic, except for the
`FenceSignalFuture` which would block for the fence. A `#[must_use]` attribute is probably enough
for this not to be a problem. When Rust gets linear types, there won't be any risk of panic
anymore.

Adding an operation after a future consumes it and produces a new future that contains both the
operation to perform and the future that was consumed. The user is expected to break the chain
by creating a fence from time to time. This is also conformant to what Vulkan experts say are
good practices.

# Queue chain and queue transfers

Since queues execute in parallel, when we submit a command buffer after a previous command buffer
they must both be executed on the same queue.

The author of this RFC thinks that it is not worth splitting the world between 
futures-that-let-you-change-queue and futures-that-dont-let-you-change-queue.

The `exec_after` function should let you specify a queue, but will potentially result in a runtime
error. If this runtime error happens to be problematic in practice, this can always be changed
later. A `exec_after_same_queue` function could be provided for convenience.

In order to be consistent with this new explicit design, changing the queue family of a resource
in non-shared mode at the end of a command buffer should be explicitly done by the user and should
also be checked when querying a future whether a resource is accessible.

Only two things would remain implicit in the submission system:

- Adding pipeline barriers between subsequent command buffers.
- "Receiving" a resource from another queue family (implicit pipeline barrier).

# Misc: events

Until now vulkano hasn't taken Vulkan events into account. They were just ignored and treated as
something to implement later.

But with this design one can easily imagine having a struct `EventSignaledFuture` that implements
`GpuFuture`, and the action of waiting on an event would take ownership of this
`EventSignaledFuture`.

The design for events is still blurry, but at least it would be a first step.

# Misc: sparse binding operations

It will be very easy in the future to integrate sparse binding operations in this design.
Performing a sparse binding would return a future that can be integrated in a chain.

# Detailed design

```rust
unsafe trait GpuFuture {
    fn is_submitted(&self) -> bool;

    fn flush(&mut self) -> Result<(), ...>;

    fn subsequent_queue_allowed(&self, queue: &Queue) -> bool;

    fn check_buffer_access(&self, &Buffer, exclusive: bool, queue: &Queue) -> bool;

    fn check_image_access(&self, &Image, exclusive: bool, queue: &Queue) -> bool;

    fn then_exec<Cb>(self, queue: &Queue, command_buffer: Cb) -> CommandBufferExecFuture<Self, Cb> {
        ...
    }

    fn then_signal_semaphore(self) -> SemaphoreSignalFuture<Self> {
        ...
    }

    ... multiple other then_* methods ...
}

unsafe impl GpuFuture for Box<GpuFuture> { ... }
```

```rust
unsafe trait Buffer {
    ... existing methods ...

    fn gpu_access(&self, exclusive: bool) -> bool;
}

// Typical implementation
unsafe impl Buffer for MyBuffer {
    #[inline]
    fn gpu_access(&self, _: bool) -> bool {
        !self.gpu_accessed.swap(true, Ordering::SeqCst)
    }
}

unsafe trait Image {
    ... existing methods ...

    fn gpu_access(&self, exclusive: bool) -> bool;
}
```

```rust
impl Swapchain {
    pub fn acquire_next_image(timeout: Duration) -> Result<SwapchainAcquireFuture, AcquireError> {
        ...
    }
}
```

**Comment by tomaka** [comment]
I'm trying to implement this, but I'm having trouble implementing submitting a chain of futures.
This requires exposing some API on GpuFuture that I haven't talked about in the RFC because I thought it was trivial. Turns out it's not.
It's not a big problem, but I feel a bit stupid for not finding a good design for this.

**Comment by tomaka** [comment]
`CpuAccessibleBuffer::start()` should probably be renamed `gpu_access()`, which makes more sense.

The `CpuAccessibleBuffer` struct wouldn't implement the `Buffer` trait. In fact it would only have two methods: one to read/write from the CPU, and `gpu_access()` which returns a handle that implements `Buffer`.

Calling `gpu_access()` while the function has already been called wouldn't return an error. Instead errors are only detected when submitting.

I think this API is intuitive enough. It still makes the `CpuAccessibleBuffer` are bit clumbsy to use because you must design your program to avoid using it multiple times simultaneously (just like you have to avoid calling `borrow_mut()` on a RefCell if it is already borrowed). It would still ultimately be replaced with upload and download pools.

**Comment by tomaka** [comment]
More feedback when trying to implement this:

- If the user executes two command buffers and joins the futures on two different queues, there's no way to do that correctly and an error must be returned.
- We need to split into two different future types when a command buffer is executed with or without a previous future.
- There are lots of side-effects to the change. Many parts of the code are no longer needed (like the semaphores queue in the swapchain). The Fence::new and Semaphore::new functions no longer need to return Arcs.
- I'm still unsure how to handle specifying the pipeline stages when waiting for a semaphore.

**Comment by tomaka** [comment]
The idea of panicking in the future's destructor if it wasn't submitted is in fact bad:

- If the future's destructor is called while unwinding, you have a double panic and the program stops.
- Since a future holds the previous futures, if a future's destructor panics then it is very likely that the previous future's destructor will panic as well and we have a double panic.
- If the user decides to use `catch_unwind` then we are unsound. Rust's policy is that all safe code must be exception-safe.

Instead maybe the solution is to make a future's destructor submit a fence and wait for it to be signaled.

> We need to split into two different future types the two situations when a command buffer is executed with or without a previous future.

I added a `DummyFuture` type instead, so that we only need one type for command buffer futures.

**Comment by tomaka** [comment]
Merged in the `incoming` branch.

**Comment by knappador** [comment]
I feel so much pain in this post :crying_cat_face:
